catalog/rocks/revparse.go

```diff
@@ -0,0 +1,76 @@
+package rocks
+
+import (
+	"regexp"
+	"strconv"
+)
+
+var (
+	hashRegexp      = regexp.MustCompile("^[a-fA-F0-9]{1,64}$")
+	modifiersRegexp = regexp.MustCompile("(^|[~^])[^^~]*")
+)
+
+type RevModType uint8
+
+const (
+	RevModTypeTilde RevModType = iota
+	RevModTypeCaret
+)
+
+func isAHash(part string) bool {
+	return hashRegexp.MatchString(part)
+}
+
+type ParsedRev struct {
+	BaseRev   string
+	Modifiers []RevModifier
+}
+
+type RevModifier struct {
+	Type  RevModType
+	Value int
+}
+
+func parseMod(buf string) (RevModifier, error) {
+	amount := 1
+	var err error
+	if len(buf) > 1 {
+		amount, err = strconv.Atoi(buf[1:])
+		if err != nil {
+			return RevModifier{}, ErrInvalidRef
+		}
+	}
+	var typ RevModType
+	switch buf[0] {
+	case '~':
+		typ = RevModTypeTilde
+	case '^':
+		typ = RevModTypeCaret
+	default:
+		return RevModifier{}, ErrInvalidRef
+	}
+
+	return RevModifier{
+		Type:  typ,
+		Value: amount,
+	}, nil
+}
+
+func RevParse(r Ref) (ParsedRev, error) {
+	p := ParsedRev{
+		Modifiers: make([]RevModifier, 0),
+	}
+	parts := modifiersRegexp.FindAllString(string(r), -1)
+	p.BaseRev = parts[0]
+	if p.BaseRev == "" {
+		return p, ErrInvalidRef
+	}
+	for _, part := range parts[1:] {
```

Comment:
This is really hard to understand.  You use the fact that all the matches must occur consecutively because of the structure of the regexp.  So now I don't understand why you allow the string to start without `^` or `~`.

I think this would really be more readable as a loop without regexps:
```go
for len(r) > 0 {
  typ := ParseType(r[0])
  if typ == "" { error... }
  r = r[0:]
  value, r := parseZeroOrMoreDigits(r)
  if value == 0 { value = 1 }
  p.Modifiers = append(p.Modifiers, RevModifier{type, value})
}
```
where `parseZeroOrMoreDigits` eats as many digits as it can off of `r` and returns the digits as an integer and the rest of r.

Reply:
that's funny because I originally had pretty much that. I accepted @johnnyaug 's suggestion that changed it into a regexp to make it easier to understand :) Given how subjective "understandability" seems to be, in regards to regexp in particular, I'm reluctant to make another change unless we have very good reasoning to do so.